server/index-node/src/schema.graphql

```diff
@@ -57,6 +59,21 @@ type EthereumIndexingStatus implements ChainIndexingStatus {
   lastHealthyBlock: Block
 }
 
+type EntityChanges {
+  updates: [EntityData!]!
+  deletes: [String!]!
+}
+
+type EntityData {
+  name: String!
+  fields: [EntityField!]!
+}
+
+type EntityField {
+  field: String!
+  valueAsJSON: String!
+}
+
```

Comment:
Thinking about the structure of the response a little more, I think we'd want something like
```
{
  "data": {
     "updates": {
        "EntityType": [
            { "id": "..", "attr1": "..", .. },
            { "id": "..", "attr1": "..", .. },
            ..
         ],
        ..
     },
     "deletes": {
        "EntityType": [ "id1", "id2", ... ],
        ..
     }
  }
}
```

With that, attribute values could be serialized with the normal rules for `r::Value`. That would require creating a custom `scalar JSON` type in the GraphQL schema (I think that would just be a declaration)

A lot of this hinges on how this data will be consumed; the other option is to return each entity as a string that is a JSON blob with the entity's data in it, but that's a lot less friendly to consume for clients since they have to parse each entity twice. @Jannis how do you want to consume the responses here?

Reply:
Yup, JSON blobs are not as user-friendly. I'm not familiar with GraphQL custom scalars, but if it doesn't require lots of extra code and we can reuse `r::Value` I can easily do that.

Reply:
GraphQL strongly types the keys in objects like `"updates"`, so we can't use arbitrary entity type names as keys — unless we make the schema something like

```graphql
type EntityChanges {
  updates: JSONObject
  deletes: JSONObject
}
```
which is almost like using an `Any` type of sorts. The best I think we can do is something like this:

```graphql
type EntityChanges! {
  updates: [EntityTypeUpdates!]!
  deletions: [EntityTypeDeletions!]!
}

type EntityTypeUpdates {
  type: String!
  entities: [JSONObject!]!
}

type EntityTypeDeletions {
  type: String!
  entities: [ID!]!
}
```

which would return something like

```json
{
  "data": {
     "updates": [
        {
          "type": "EntityType",
          "entities": [
            { "id": "..", "attr1": "..", .. },
            { "id": "..", "attr1": "..", .. },
            ..
          ],
          }
        ..
     ],
     "deletes": {
        { "type": "EntityType", "entities": [ "id1", "id2", .. ] },
        ..
     }
  }
}
```

---

If we wanted the result to be even more specific (wrt types), we could replace the `JSONObject` type with something like

```graphql
type ChangedEntity {
  type: String!
  fields: [ChangedEntityField!]!
}

type ChangedEntityField {
  name: String!
  value: ChangedValue!
}

interface ChangedValue {
  type: String!
}

type ChangedBigIntValue implements ChangedValue {
  type: String!
  value: BigInt!
}

type ChangedStringValue implements ChangedValue {
  type: String!
  value: String
}

type ChangedListValue implements ChangedValue {
  type: String!
  value: [ChangedValue!]
}

# Repeat for all possible value types.
```

But I think that's overkill. I'd be ok with entities being represented with `scalar JSON`, which we should just be able to add at the top of the GraphQL schema.

Reply:
Having said all this, I like Filippo's proposed schema with fields in a list, having a `name: String!` and `value: JSON` field. I think that's a good compromise because the response structure is typed down to the field values — which could really be anything.

Reply:
To me `Vec<{ key: String, value: Any }>` and `Map<String, Any>` are more or less the same. I can easily build a map from a list of key/value pairs returned by GraphQL on the other side.

Reply:
Having the map of entities themselves as a JSON blob is quite bad and we lose a lot of GraphQL type information, I agree. That's why I'm listing entities in a `r::List`. Fields themselves can and should use JSON to not complicate things unnecessarily though :+1:

Reply:
I still don't understand why we expect multiple entities with the same entity type to be returned from this query.

Reply:
We don't. But then we don't need to check the two sets for collisions, right?

Reply:
I settled on this API:

```
type EntityChanges! {
  updates: [EntityTypeUpdates!]!
  deletions: [EntityTypeDeletions!]!
}

type EntityTypeUpdates {
  type: String!
  entities: [JSONObject!]!
}

type EntityTypeDeletions {
  type: String!
  entities: [ID!]!
}
```

Marking this as resolved and continuing the discussion about update/removal cross-checking to another comment.